log/value.go

```diff
@@ -0,0 +1,68 @@
+package log
+
+import (
+	"time"
+
+	"gopkg.in/stack.v1"
+)
+
+// BindValues returns a slice with all value elements (odd indexes) that
+// implement Valuer replaced with the result of calling their Value method. If
+// no value elements implement Valuer, the original slice is returned.
+func BindValues(keyvals ...interface{}) []interface{} {
+	if !containsValuer(keyvals) {
+		return keyvals
+	}
+
+	bound := make([]interface{}, len(keyvals))
+	copy(bound, keyvals)
+	for i := 1; i < len(bound); i += 2 {
+		if v, ok := bound[i].(Valuer); ok {
+			bound[i] = v.Value()
+		}
+	}
+
+	return bound
+}
+
+func containsValuer(keyvals []interface{}) bool {
+	for i := 1; i < len(keyvals); i += 2 {
+		if _, ok := keyvals[i].(Valuer); ok {
+			return true
+		}
+	}
+	return false
+}
+
+// Value is the return type of the Value method of the Valuer interface.
+type Value interface{}
+
+// A Valuer is able to convert itself into log Value.
+//
+// A Valuer passed to With stores a dynamic value which is reevaluated on each
+// log event.
+type Valuer interface {
+	// Value returns a log Value instead of an interface{} to avoid
+	// inadvertently matching types from packages not intended for use with
+	// the gokit/log package.
+	Value() Value
+}
+
+type timeStamp struct{}
```

Comment:
Mutating global state has its drawbacks. For instance, running parallel
tests that need Timestamp would be harder.

On Friday, April 24, 2015, Chris Hines <notifications@github.com> wrote:

> In log/value.go
> <https://github.com/peterbourgon/gokit/pull/32#discussion_r29016380>:
>
> > +
> > +// Value is the return type of the Value method of the Valuer interface.
> > +type Value interface{}
> > +
> > +// A Valuer is able to convert itself into log Value.
> > +//
> > +// A Valuer passed to With stores a dynamic value which is reevaluated on each
> > +// log event.
> > +type Valuer interface {
> > +	// Value returns a log Value instead of an interface{} to avoid
> > +	// inadvertently matching types from packages not intended for use with
> > +	// the gokit/log package.
> > +	Value() Value
> > +}
> > +
> > +type timeStamp struct{}
>
> I don't have a strong opinion about ValuerFunc.
>
> For testing purposes would it be sufficient to just assign a new value to
> log.Timestamp?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/peterbourgon/gokit/pull/32/files#r29016380>.
>


-- 
Sent from Gmail Mobile